shared/src/current_block.rs

```diff
@@ -0,0 +1,100 @@
+use anyhow::{anyhow, Context, Result};
+use futures::Stream;
+use primitive_types::H256;
+use std::{future::Future, time::Duration};
+use tokio::sync::watch;
+use web3::{
+    types::{BlockId, BlockNumber},
+    Transport, Web3,
+};
+
+pub type Block = web3::types::Block<H256>;
+
+const POLL_INTERVAL: Duration = Duration::from_secs(1);
+
+/// Creates a clonable stream that yields the current block whenever it changes and a future
+/// driving the updates of current block.
+///
+/// The stream is not guaranteed to yield *every* block individually without gaps but it does yield
+/// the newest block whenever it changes. In practice this means that if the node changes the
+/// current block in quick succession we might only observe the last block, skipping some blocks in
+/// between. If the the current block has not yet been fetched the first time the stream yields a
+/// default block with hash and number 0.
+///
+/// The future runs until all associated streams have been dropped.
+pub fn current_block_stream(
+    web3: Web3<impl Transport>,
+) -> (
+    impl Stream<Item = Block> + Clone + Unpin,
+    impl Future<Output = ()>,
+) {
+    let default_hash = H256::default();
+    let default_block = Block {
+        hash: Some(default_hash),
+        number: Some(0.into()),
+        ..Default::default()
+    };
+    let (sender, receiver) = watch::channel(default_block);
```

Comment:
I find yielding an "invalid" default block initially really hacky. Could we instead pass `Option<Block>` and wrap receiver in a `filter_map` that only forwards the `Some` items?

Reply:
Agree. Implemented this in the first new commit. Had to introduce a wrapper type because filter_map's Stream isn't clonable.